File: cogs/logging.py
"""
Logging Cog
Logs to MOD_LOG_CHANNEL:
  - Member join / leave
  - Message edit / delete
  - Member ban / unban
"""
import discord
from discord.ext import commands
from utils.utils import log_action, utcnow

class Logging(commands.Cog):
    def __init__(self, bot):
        self.bot = bot

    @commands.Cog.listener()
    async def on_member_join(self, member: discord.Member):
        embed = discord.Embed(
            title="📥 Member Joined",
            color=discord.Color.green(),
            timestamp=utcnow()
        )
        embed.set_thumbnail(url=member.display_avatar.url)
        embed.add_field(name="User", value=f"{member} (`{member.id}`)")
        embed.add_field(name="Account Created", value=discord.utils.format_dt(member.created_at, 'R'))
        embed.set_footer(text=f"Member count: {member.guild.member_count}")
        await log_action(member.guild, embed)

    @commands.Cog.listener()
    async def on_member_remove(self, member: discord.Member):
        embed = discord.Embed(
            title="📤 Member Left",
            color=discord.Color.greyple(),
            timestamp=utcnow()
        )
        embed.add_field(name="User", value=f"{member} (`{member.id}`)")
        roles = [r.mention for r in member.roles if r.name != "@everyone"]
        embed.add_field(name="Roles", value=", ".join(roles) or "None")
        await log_action(member.guild, embed)

    @commands.Cog.listener()
    async def on_message_edit(self, before: discord.Message, after: discord.Message):
        if not before.guild or before.author.bot:
            return
        if before.content == after.content:
            return
        embed = discord.Embed(
            title="✏️ Message Edited",
            color=discord.Color.blue(),
            timestamp=utcnow()
        )
        embed.add_field(name="Author", value=f"{before.author} (`{before.author.id}`)")
        embed.add_field(name="Channel", value=before.channel.mention)
        embed.add_field(name="Before", value=before.content[:1024] or "*empty*", inline=False)
        embed.add_field(name="After",  value=after.content[:1024] or "*empty*", inline=False)
        embed.add_field(name="Jump", value=f"[Link]({after.jump_url})")
        await log_action(before.guild, embed)

    @commands.Cog.listener()
    async def on_message_delete(self, message: discord.Message):
        if not message.guild or message.author.bot:
            return
        embed = discord.Embed(
            title="🗑️ Message Deleted",
            color=discord.Color.red(),
            timestamp=utcnow()
        )
        embed.add_field(name="Author",  value=f"{message.author} (`{message.author.id}`)")
        embed.add_field(name="Channel", value=message.channel.mention)
        embed.add_field(name="Content", value=message.content[:1024] or "*empty/attachment*", inline=False)
        await log_action(message.guild, embed)

    @commands.Cog.listener()
    async def on_member_ban(self, guild: discord.Guild, user: discord.User):
        embed = discord.Embed(title="🔨 Member Banned", color=discord.Color.dark_red(), timestamp=utcnow())
        embed.add_field(name="User", value=f"{user} (`{user.id}`)")
        await log_action(guild, embed)

    @commands.Cog.listener()
    async def on_member_unban(self, guild: discord.Guild, user: discord.User):
        embed = discord.Embed(title="✅ Member Unbanned", color=discord.Color.green(), timestamp=utcnow())
        embed.add_field(name="User", value=f"{user} (`{user.id}`)")
        await log_action(guild, embed)

async def setup(bot):
    await bot.add_cog(Logging(bot))

File: src/handlers/logging.rs
use serenity::all::*;
use chrono::Utc;
use crate::{BotData, utils::log_action};

fn utcnow() -> Timestamp {
    Timestamp::from(Utc::now())
}

/// on_guild_member_addition
pub async fn handle_member_join(ctx: &Context, member: &Member) {
    let mod_log_channel = {
        let data = ctx.data.read().await;
        data.get::<BotData>().unwrap().config.mod_log_channel
    };

    let account_created = format!("<t:{}:R>", member.user.created_at().unix_timestamp());

    // Get member count from cache (no await needed)
    let member_count = ctx.cache
        .guild(member.guild_id)
        .map(|g| g.member_count)
        .unwrap_or(0);

    let avatar_url = member.user.face();
    let user_tag   = member.user.tag();
    let user_id    = member.user.id;

    let embed = CreateEmbed::new()
        .title("📥 Member Joined")
        .color(Color::DARK_GREEN)
        .timestamp(utcnow())
        .thumbnail(avatar_url)
        .field("User",            format!("{user_tag} (`{user_id}`)"), false)
        .field("Account Created", account_created, false)
        .footer(CreateEmbedFooter::new(format!("Member count: {member_count}")));

    log_action(&ctx.http, mod_log_channel, embed).await;
}

/// on_guild_member_removal
pub async fn handle_member_leave(ctx: &Context, _guild_id: GuildId, user: &User, member_data: &Option<Member>) {
    let mod_log_channel = {
        let data = ctx.data.read().await;
        data.get::<BotData>().unwrap().config.mod_log_channel
    };

    let roles_str = member_data
        .as_ref()
        .map(|m| {
            let r: Vec<String> = m.roles.iter().map(|rid| format!("<@&{}>", rid)).collect();
            if r.is_empty() { "None".to_string() } else { r.join(", ") }
        })
        .unwrap_or_else(|| "Unknown".to_string());

    let embed = CreateEmbed::new()
        .title("📤 Member Left")
        .color(Color::LIGHT_GREY)
        .timestamp(utcnow())
        .field("User",  format!("{} (`{}`)", user.tag(), user.id), false)
        .field("Roles", roles_str, false);

    log_action(&ctx.http, mod_log_channel, embed).await;
}

/// on_message_update
pub async fn handle_message_edit(
    ctx:    &Context,
    old:    &Option<Message>,
    new:    &Option<Message>,
    _event: &MessageUpdateEvent,
) {
    let (before, after) = match (old, new) {
        (Some(b), Some(a)) => (b, a),
        _ => return,
    };
    if before.author.bot || before.guild_id.is_none() {
        return;
    }
    if before.content == after.content {
        return;
    }

    let mod_log_channel = {
        let data = ctx.data.read().await;
        data.get::<BotData>().unwrap().config.mod_log_channel
    };

    let embed = CreateEmbed::new()
        .title("✏️ Message Edited")
        .color(Color::BLUE)
        .timestamp(utcnow())
        .field("Author",  format!("{} (`{}`)", before.author.tag(), before.author.id), false)
        .field("Channel", format!("<#{}>", before.channel_id), false)
        .field("Before",  truncate(&before.content, 1024), false)
        .field("After",   truncate(&after.content, 1024), false)
        .field("Jump",    format!("[Link]({})", after.link()), false);

    log_action(&ctx.http, mod_log_channel, embed).await;
}

/// on_message_delete
pub async fn handle_message_delete(
    ctx:        &Context,
    channel_id: ChannelId,
    deleted_id: MessageId,
    guild_id:   Option<GuildId>,
) {
    if guild_id.is_none() {
        return;
    }

    let mod_log_channel = {
        let data = ctx.data.read().await;
        data.get::<BotData>().unwrap().config.mod_log_channel
    };

    // Best effort — message may not be in cache
    let (author_str, content_str) = ctx.cache
        .message(channel_id, deleted_id)
        .filter(|msg| !msg.author.bot)
        .map(|msg| (
            format!("{} (`{}`)", msg.author.tag(), msg.author.id),
            truncate(&msg.content, 1024),
        ))
        .unwrap_or_else(|| (
            format!("Unknown (msg `{deleted_id}`)"),
            "*not in cache*".to_string(),
        ));

    let embed = CreateEmbed::new()
        .title("🗑️ Message Deleted")
        .color(Color::RED)
        .timestamp(utcnow())
        .field("Author",  author_str, false)
        .field("Channel", format!("<#{channel_id}>"), false)
        .field("Content", content_str, false);

    log_action(&ctx.http, mod_log_channel, embed).await;
}

/// on_guild_ban_addition
pub async fn handle_ban(ctx: &Context, _guild_id: GuildId, user: &User) {
    let mod_log_channel = {
        let data = ctx.data.read().await;
        data.get::<BotData>().unwrap().config.mod_log_channel
    };

    let embed = CreateEmbed::new()
        .title("🔨 Member Banned")
        .color(Color::DARK_RED)
        .timestamp(utcnow())
        .field("User", format!("{} (`{}`)", user.tag(), user.id), false);

    log_action(&ctx.http, mod_log_channel, embed).await;
}

/// on_guild_ban_removal
pub async fn handle_unban(ctx: &Context, _guild_id: GuildId, user: &User) {
    let mod_log_channel = {
        let data = ctx.data.read().await;
        data.get::<BotData>().unwrap().config.mod_log_channel
    };

    let embed = CreateEmbed::new()
        .title("✅ Member Unbanned")
        .color(Color::DARK_GREEN)
        .timestamp(utcnow())
        .field("User", format!("{} (`{}`)", user.tag(), user.id), false);

    log_action(&ctx.http, mod_log_channel, embed).await;
}

fn truncate(s: &str, max: usize) -> String {
    if s.len() <= max {
        s.to_string()
    } else {
        format!("{}…", &s[..max.saturating_sub(1)])
    }
}
